integrations/micronaut/cdi/src/main/java/io/helidon/integrations/micronaut/cdi/MicronautMethodInvocationContext.java

```diff
@@ -0,0 +1,221 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.integrations.micronaut.cdi;
+
+import java.lang.reflect.Method;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import javax.interceptor.InvocationContext;
+
+import io.micronaut.aop.Interceptor;
+import io.micronaut.aop.MethodInterceptor;
+import io.micronaut.aop.MethodInvocationContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.convert.value.MutableConvertibleMultiValuesMap;
+import io.micronaut.core.convert.value.MutableConvertibleValues;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.ArgumentValue;
+import io.micronaut.core.type.MutableArgumentValue;
+import io.micronaut.core.type.ReturnType;
+import io.micronaut.inject.ExecutableMethod;
+
+/**
+ * Invocation context for Micronaut interceptors.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+class MicronautMethodInvocationContext implements MethodInvocationContext {
+    private static final Logger LOGGER = Logger.getLogger(MicronautMethodInvocationContext.class.getName());
+
+    private final InvocationContext cdiContext;
+    private final ExecutableMethod executableMethod;
+    private final Set<MethodInterceptor<?, ?>> allInterceptors;
+    private final MutableConvertibleValues attributes;
+    private final Map<String, MutableArgumentValue<?>> mutableArguments;
+
+    private Iterator<MethodInterceptor<?, ?>> remaining;
+
+    private MicronautMethodInvocationContext(InvocationContext cdiContext,
+                                             ExecutableMethod executableMethod,
+                                             Set<MethodInterceptor<?, ?>> allInterceptors,
+                                             Iterator<MethodInterceptor<?, ?>> remaining) {
+        this.cdiContext = cdiContext;
+        this.executableMethod = executableMethod;
+        this.allInterceptors = allInterceptors;
+        this.remaining = remaining;
+
+        this.attributes = new MutableConvertibleMultiValuesMap();
+        this.mutableArguments = new LinkedHashMap<>();
+        Object[] parameters = cdiContext.getParameters();
+        Argument[] argumentNames = executableMethod.getArguments();
+
+        for (int i = 0; i < argumentNames.length; i++) {
+            Argument argument = argumentNames[i];
+            Object parameterValue = parameters[i];
+            mutableArguments.put(argument.getName(),
+                                 new MutableArgument<Object>(argument, parameterValue));
+        }
```

Comment:
In Micronaut core we construct the attributes and mutable arguments lazily since they may not be used which reduces overhead